### PR TITLE
OP-319 Added the ability to select package dimensions

### PR DESCRIPTION
**Pull request by tomkalon**
- Added the ability to select package dimensions

Changing the dimensions of the package is possible until the export shipment is sent.

Przechwytywanie

### CODE REVIEW

File: features/changing_shipping_method_parcel_template.feature [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    And those orders were placed with "Inpost" shipping method
+    And set product weight to "10"
+    And set units to the shipment
+
+  @ui
+  Scenario: Seeing shipments to export
+    When I go to the shipping export page
+    Then I should see 1 shipments with "New" state
+    When I select parcel template
+    Then I should see that shipping export parcel template is set
```
Comment (senghe) [inline features/changing_shipping_method_parcel_template.feature:28]: This seem to be two scenarios. I can approve this, as the scenario is not so complex, but the order of keywords should be at most: Given/When/Then and after then cannot be next Given/When.

File: spec/EventListener/SelectParcelTemplateEventListenerSpec.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        $this
+            ->shouldThrow(\Exception::class)
+            ->during('setParcelTemplate', [$event])
+        ;
+    }
+
+    /**
+     * @throws \Exception
+     */
```
Comment (senghe) [inline spec/EventListener/SelectParcelTemplateEventListenerSpec.php:69]: We do not put @throws annotation.
```suggestion
```

File: src/Api/WebClient.php [hunk 3]
```diff
@@ -149,8 +152,12 @@ public function getShipments(): ?array
         return $this->request('GET', $url);
     }
 
-    public function createShipment(ShipmentInterface $shipment): array
-    {
+    public function createShipment(
+        ShipmentInterface $shipment,
+        ?ShippingExportInterface $shippingExport = null,
+    ): array {
+        $this->shippingExport = $shippingExport;
```
Comment (senghe) [inline src/Api/WebClient.php:159]: There is no need for putting this as a class property. You can put this into the dependent methods as a parameter.

File: src/Api/WebClient.php [hunk 4]
```diff
@@ -317,9 +324,14 @@ private function isCashOnDelivery(OrderInterface $order): bool
         return false;
     }
 
-    private function createParcel(ShipmentInterface $shipment): array
-    {
+    private function createParcel(
+        ShipmentInterface $shipment,
+    ): array {
         $weight = $shipment->getShippingWeight();
+        $template = 'large';
```
Comment (senghe) [inline src/Api/WebClient.php:331]: I think we should put the value into container parameter, as it should be configurable.

File: src/Api/WebClientInterface.php [hunk 5]
```diff
@@ -78,7 +79,7 @@ public function getLabels(array $shipmentIds): ?string;
 
     public function getAuthorizedHeaderWithContentType(): array;
 
-    public function createShipment(ShipmentInterface $shipment): array;
+    public function createShipment(ShipmentInterface $shipment, ?ShippingExportInterface $shippingExport = null): array;
```
Comment (senghe) [inline src/Api/WebClientInterface.php:82]: Why there can be null?

File: src/Resources/config/grid/order.yml [hunk 6]
```diff
@@ -2,33 +2,85 @@
...
-                  number: resource.number
-          pay:
-            type: shop_pay
-            label: sylius.ui.pay
-            options:
-              link:
-                route: sylius_shop_order_show
-                parameters:
-                  tokenValue: resource.tokenValue
+    templates:
```
Comment (senghe) [inline src/Resources/config/grid/order.yml:5]: What's the reason of so many lines changed? I can't read what exactly was the logic change...

File: src/Resources/config/routes.yml [hunk 7]
```diff
@@ -8,3 +8,9 @@ bitbag_sylius_inpost_plugin_admin_add_point_to_order:
     methods: [POST]
     defaults:
         _controller: bitbag.sylius_inpost_plugin.controller.add_point_to_order::addPointToOrderAction
+    
+bitbag_admin_shipping_export_select_parcel_template:
+    path: /admin/shipping_export/select_parcel_template/{id}/{template}
+    methods: [POST, PUT]
+    defaults:
+        _controller: bitbag.controller.shipping_export::selectParcelTemplate
```
Comment (senghe) [inline src/Resources/config/routes.yml:16]: I could create separate controller for this with __invoke() method.

File: tests/Application/config/packages/bitbag_sylius_shipping_export_plugin.yaml [hunk 8]
```diff
@@ -1,2 +1,9 @@
...
-    - { resource: "@BitBagSyliusShippingExportPlugin/Resources/config/config.yml" }
\ No newline at end of file
+    - { resource: "@BitBagSyliusShippingExportPlugin/Resources/config/config.yml" }
+
+sylius_resource:
+    resources:
+        bitbag.shipping_export:
+            classes:
+                model: Tests\BitBag\SyliusInPostPlugin\Application\src\Entity\ShippingExport
+                controller: BitBag\SyliusInPostPlugin\Controller\ShippingExportController
```
Comment (senghe) [inline tests/Application/config/packages/bitbag_sylius_shipping_export_plugin.yaml:9]: Is this really needed? Do you override anything in ShippingExportController::exportSingleShipmentAction() or ShippingExportController::exportAllNewShipmentsAction() ? If yes, couldn't it be changed in any other place in the business logic?

**Review by senghe — CHANGES_REQUESTED** [review]

File: features/changing_shipping_method_parcel_template.feature (hunk 1 above)
Comment (tomkalon) [reply features/changing_shipping_method_parcel_template.feature:28]: And a better solution would not be flipping: "Scenario: Seeing shipments to export" to the: 
@managing_shipping_export_inpost?

The same background.

File: spec/EventListener/SelectParcelTemplateEventListenerSpec.php (hunk 2 above)
Comment (tomkalon) [reply spec/EventListener/SelectParcelTemplateEventListenerSpec.php:69]: OK 👍

File: src/Api/WebClient.php (hunk 3 above)
Comment (tomkalon) [reply src/Api/WebClient.php:159]: Done 👍

File: src/Api/WebClient.php (hunk 4 above)
Comment (tomkalon) [reply src/Api/WebClient.php]: Done 👍 
Added ability to select default parcel_template and label_type to .env file.

File: src/Api/WebClientInterface.php (hunk 5 above)
Comment (tomkalon) [reply src/Api/WebClientInterface.php]: Changed 👍 
I took too timid an approach.

File: src/Resources/config/grid/order.yml (hunk 6 above)
Comment (tomkalon) [reply src/Resources/config/grid/order.yml:5]: Ctrl+ALt+L and changed for 4 spaces as tab.
My fault. I will remember for the future.

File: src/Resources/config/routes.yml (hunk 7 above)
Comment (tomkalon) [reply src/Resources/config/routes.yml:16]: N/A 👍

File: doc/installation.md [hunk 9]
```diff
@@ -37,8 +37,138 @@ bitbag_sylius_inpost_plugin:
...
+use BitBag\SyliusShippingExportPlugin\Event\ExportShipmentEvent;
+use BitBag\SyliusShippingExportPlugin\Repository\ShippingExportRepositoryInterface;
+use Sylius\Bundle\ResourceBundle\Controller\ResourceController;
+use Sylius\Component\Resource\Model\ResourceInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Webmozart\Assert\Assert;
+
+final class ShippingExportController extends ResourceController
```
Comment (senghe) [inline doc/installation.md:66]: I would extend namespace `BitBag\SyliusShippingExportPlugin\Controller\ShippingExportController` and remove the public methods from the listing.

File: doc/installation.md [hunk 10]
```diff
@@ -37,8 +37,138 @@ bitbag_sylius_inpost_plugin:
...
+        order:
+            classes:
+                model: App\Entity\Order
+
+sylius_shipping:
+    resources:
+        shipping_method:
+            classes:
+                model: App\Entity\ShippingMethod
+
```
Comment (senghe) [inline doc/installation.md:170]:
```suggestion

```

File: doc/installation.md [hunk 11]
```diff
@@ -105,28 +260,271 @@ prefer)
...
+        assets:
+            packages:
+                admin:
+                    json_manifest_path: '%kernel.project_dir%/public/build/admin/manifest.json'
+                shop:
+                    json_manifest_path: '%kernel.project_dir%/public/build/shop/manifest.json'
+                app.admin:
+                    json_manifest_path: '%kernel.project_dir%/public/build/app/admin/manifest.json'
+                app.shop:
+                    json_manifest_path: '%kernel.project_dir%/public/build/app/shop/manifest.json'
```
Comment (senghe) [inline doc/installation.md:495]:
```suggestion
// ...
```

File: doc/installation.md [hunk 12]
```diff
@@ -105,28 +260,271 @@ prefer)
...
+1. Before Webpack installation, please create the `config/packages/webpack_encore.yaml` file with a content of:
+
+    ```yaml
+    webpack_encore:
+        output_path: '%kernel.project_dir%/public/build/default'
+        builds:
+            admin: '%kernel.project_dir%/public/build/admin'
+            shop: '%kernel.project_dir%/public/build/shop'
+            app.admin: '%kernel.project_dir%/public/build/app/admin'
+            app.shop: '%kernel.project_dir%/public/build/app/shop'
```
Comment (senghe) [inline doc/installation.md:447]:
```suggestion
// ...
```

File: spec/Api/WebClientSpec.php [hunk 13]
```diff
@@ -26,9 +28,16 @@ final class WebClientSpec extends ObjectBehavior
 
     public const LABEL_TYPE = "normal";
 
-    public function let(Client $client): void
+    public const PARCEL_TEMPLATE = "medium";
+
+
```
Comment (senghe) [inline spec/Api/WebClientSpec.php:33]:
```suggestion

```

File: tests/Application/src/Controller/ShippingExportController.php [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+use BitBag\SyliusShippingExportPlugin\Event\ExportShipmentEvent;
+use BitBag\SyliusShippingExportPlugin\Repository\ShippingExportRepositoryInterface;
+use Sylius\Bundle\ResourceBundle\Controller\ResourceController;
+use Sylius\Component\Resource\Model\ResourceInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Webmozart\Assert\Assert;
+
+final class ShippingExportController extends ResourceController
```
Comment (senghe) [inline tests/Application/src/Controller/ShippingExportController.php:24]: Please do this as I asked you in README.md comment

**Review by senghe — APPROVED** [review]
I'm approving in advance. Please do the thing I asked you with ResourceController and you're free :vulcan_salute:

File: doc/installation.md (hunk 9 above)
Comment (tomkalon) [reply doc/installation.md:66]: I can't extend ShippingExportController because it is a final class.

File: doc/installation.md (hunk 10 above)
Comment (tomkalon) [reply doc/installation.md]: Fixed... and few other similar more.

File: doc/installation.md (hunk 11 above)
Comment (tomkalon) [reply doc/installation.md]: Fixed.

File: doc/installation.md (hunk 12 above)
Comment (tomkalon) [reply doc/installation.md]: Fixed.

File: spec/Api/WebClientSpec.php (hunk 13 above)
Comment (tomkalon) [reply spec/Api/WebClientSpec.php]: Fixed.

**Pull request merged**